src/server/scripts/EasternKingdoms/zone_silverpine_forest.cpp

```diff
@@ -15,125 +15,5311 @@
-SDCategory: Silverpine Forest
-EndScriptData */
 
-/* ContentData
-npc_deathstalker_erland
-EndContentData */
+#include "CombatAI.h"
+#include "CreatureAIImpl.h"
+#include "GameObject.h"
+#include "GameObjectAI.h"
+#include "GridNotifiers.h"
+#include "MotionMaster.h"
+#include "ObjectAccessor.h"
+#include "ObjectMgr.h"
+#include "Object.h"
+#include "Player.h"
+#include "ScriptMgr.h"
+#include "ScriptedCreature.h"
+#include "ScriptedGossip.h"
+#include "SharedDefines.h"
+#include "SpellAuras.h"
+#include "SpellAuraEffects.h"
+#include "SpellInfo.h"
+#include "SpellScript.h"
+#include "TemporarySummon.h"
+#include "Unit.h"
+#include "Vehicle.h"
+
+enum TalksFromCreatures
+{
+    TALK_0 = 0,
+    TALK_1,
+    TALK_2,
+    TALK_3,
+    TALK_4,
+    TALK_5,
+    TALK_6,
+    TALK_7,
+    TALK_8,
+    TALK_9,
+    TALK_10,
+    TALK_11,
+    TALK_12,
+    TALK_13,
+    TALK_14,
+    TALK_15,
+    TALK_16,
+    TALK_17,
+    TALK_18,
+    TALK_19,
+    TALK_20,
+    TALK_21,
+};
+
+enum QuestTheGilneasLiberationFront
+{
+    NPC_FORSAKEN_TROOPER1 = 44791,
+    NPC_FORSAKEN_TROOPER2 = 44792,
+    NPC_WORGEN_RENEGATE = 44793,
+
+    SPELL_HEARTSTRIKE = 84182,
+    SPELL_DARKENED = 73307,
+    SPELL_FLURRY_OF_CLAWS_CHANNEL = 80365,
+    SPELL_FLURRY_OF_CLAWS = 80367,
+
+    EVENT_SPELL_QUEUE = 102,
+    EVENT_SPELL_DAMAGE = 103,
+
+    MOVE_TO_HOMEPOSITION,
+};
+
+// Worgen Renegade - 44793
+struct npc_silverpine_worgen_renegade : public ScriptedAI
+{
+    npc_silverpine_worgen_renegade(Creature* creature) : ScriptedAI(creature) { }
+
+    void Reset() override
+    {
+        _events.Reset();
+
+        me->RemoveAura(SPELL_DARKENED);
+
+        me->SetReactState(REACT_AGGRESSIVE);
+    }
+
+    void SpellHit(Unit* caster, SpellInfo const* spell) override
+    {
+        if (!caster)
+            return;
+
+        if (spell->Id == SPELL_HEARTSTRIKE)
+            me->AddAura(SPELL_DARKENED, me);
+    }
+
+    void JustEngagedWith(Unit* /*who*/) override
+    {
+        _events.ScheduleEvent(EVENT_SPELL_QUEUE, 3s);
+    }
+
+    void UpdateAI(uint32 diff) override
+    {
+        _events.Update(diff);
+
+        while (uint32 eventId = _events.ExecuteEvent())
+        {
+            switch (eventId)
+            {
+                case EVENT_SPELL_QUEUE:
+                {
+                    DoCastVictim(SPELL_FLURRY_OF_CLAWS_CHANNEL);
+
+                    _events.ScheduleEvent(EVENT_SPELL_QUEUE, 16s, 18s);
+                    break;
+                }
+
+                default:
+                    break;
+            }
+        }
+
+        if (!UpdateVictim())
+            return;
+
+        DoMeleeAttackIfReady();
+    }
+
+private:
+    EventMap _events;
+};
+
+// Flurry of Claws - 80365
+class spell_silverpine_flurry_of_claws : public AuraScript
+{
+    void HandlePeriodic(AuraEffect const* /*aurEff*/)
+    {
+        PreventDefaultAction();
+
+        GetCaster()->CastSpell(GetTarget(), GetSpellInfo()->Effects[EFFECT_0].TriggerSpell, true);
```

Comment:
Technically `GetCaster()` can return `nullptr` here. Are you sure this will not happen for this particular case?

Reply:
I'm unsure right now that you've mentioned it. In any case, the caster should be the one triggering the effect.

Reply:
GetCaster() can be nullptr in aura scripts. a unit might have a aura while the caster went offline.

Reply:
Since it's always the caster triggering the effect, we'll keep GetTarget().